File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofStores/NCAABBettingTrendsStore.swift
import Foundation
import Observation
import Supabase
import WagerproofModels
import WagerproofServices

/// NCAAB situational betting trends store. Mirrors RN
/// `hooks/useNCAABBettingTrends.ts` end-to-end. Fetches from
/// `ncaab_game_situational_trends_today`; falls back to
/// `ncaab_game_situational_trends` when the today view is empty (matches
/// RN). After loading the rows it pairs them by `team_side`, joins logos
/// from `ncaab_team_mapping`, pulls tipoff times from `v_cbb_input_values`,
/// and computes the O/U consensus + ATS dominance scores used for sorting.
///
/// Backend queries are byte-identical to RN — the select clauses, table
/// names, and post-fetch joins must NOT diverge.
@Observable
@MainActor
public final class NCAABBettingTrendsStore {
    public enum SortMode: String, Hashable, Sendable {
        case time, ouConsensus, atsDominance
    }

    public enum LoadState: Equatable, Sendable {
        case idle, loading, loaded, refreshing
        case failed(String)
    }

    public private(set) var games: [NCAABGameTrendsData] = []
    public private(set) var loadState: LoadState = .idle
    public var sortMode: SortMode = .time {
        didSet { games = sortGames(games, mode: sortMode) }
    }
    /// Sheet-presentation state for the per-game trends detail.
    public var selectedGame: NCAABGameTrendsData?

    private let minGamesThreshold: Int = 5
    private let minPercentage: Double = 55
    private let minATSDifference: Double = 10

    public init() {}

    public func openTrendsSheet(_ game: NCAABGameTrendsData) {
        selectedGame = game
    }

    public func closeTrendsSheet() {
        selectedGame = nil
    }

    /// Lookup a single game by id. Mirrors RN's
    /// `useNCAABBettingTrendsForGame(gameId)` which reads from the same
    /// module-level cache. Returns nil if the store hasn't been refreshed or
    /// the game isn't in today's trends slate.
    public func trends(for gameId: Int) -> NCAABGameTrendsData? {
        games.first(where: { $0.gameId == gameId })
    }

    public func refresh() async {
        #if DEBUG
        // Dummy Data Mode: synthesized situational trends keyed to the captured
        // NCAAB slate so the per-game betting-trends widget populates offseason.
        if DummyDataMode.isEnabled {
            games = sortGames(DummyData.ncaabTrendsData(), mode: sortMode)
            loadState = .loaded
            return
        }
        #endif
        loadState = .loading
        let cfb = await CFBSupabase.shared.client

        // Step 1: situational trends (today's table → fallback).
        var trendsRows: [NCAABSituationalTrendRow] = []
        do {
            let primary: [NCAABSituationalTrendRow] = (try? await cfb
                .from("ncaab_game_situational_trends_today")
                .select()
                .order("game_date", ascending: true)
                .order("game_id", ascending: true)
                .execute()
                .value) ?? []
            if !primary.isEmpty {
                trendsRows = primary
            } else {
                // Fallback path mirrors RN.
                let fallback: [NCAABSituationalTrendRow] = try await cfb
                    .from("ncaab_game_situational_trends")
                    .select()
                    .order("game_date", ascending: true)
                    .order("game_id", ascending: true)
                    .execute()
                    .value
                trendsRows = fallback
            }
        } catch {
            loadState = .failed("Failed to fetch NCAAB trends: \(error.localizedDescription)")
            return
        }
        if trendsRows.isEmpty {
            games = []
            loadState = .loaded
            return
        }

        // Step 2: team logos.
        let mappingRows: [MappingRow] = (try? await cfb
            .from("ncaab_team_mapping")
            .select("api_team_id, espn_team_id")
            .execute()
            .value) ?? []
        var teamLogoMap: [Int: String?] = [:]
        for row in mappingRows {
            let espnId: Int? = {
                if let n = row.espnTeamIdInt { return n }
                if let s = row.espnTeamIdString, let parsed = Int(s) { return parsed }
                return nil
            }()
            let logo: String? = {
                guard let espnId else { return nil }
                return "https://a.espncdn.com/i/teamlogos/ncaa/500/\(espnId).png"
            }()
            teamLogoMap[row.apiTeamId] = logo
        }

        // Step 3: group rows by game_id using team_side.
        var partial: [Int: (away: NCAABSituationalTrendRow?, home: NCAABSituationalTrendRow?, awayLogo: String?, homeLogo: String?, gameDate: String)] = [:]
        for row in trendsRows {
            let side = row.teamSide.lowercased()
            guard side == "away" || side == "home" else { continue }
            let logo = teamLogoMap[row.apiTeamId] ?? nil
            var existing = partial[row.gameId] ?? (nil, nil, nil, nil, row.gameDate)
            if side == "away" {
                existing.away = row
                existing.awayLogo = logo
            } else {
                existing.home = row
                existing.homeLogo = logo
            }
            partial[row.gameId] = existing
        }

        var built: [NCAABGameTrendsData] = []
        for (gameId, slots) in partial {
            guard let away = slots.away, let home = slots.home else { continue }
            built.append(
                NCAABGameTrendsData(
                    gameId: gameId,
                    gameDate: slots.gameDate,
                    tipoffTime: nil,
                    awayTeam: away,
                    homeTeam: home,
                    awayTeamLogo: slots.awayLogo,
                    homeTeamLogo: slots.homeLogo
                )
            )
        }

        // Step 4: tipoff times from v_cbb_input_values for today's game ids.
        if !built.isEmpty {
            let gameIds = built.map { $0.gameId }
            let timeRows: [TimeRow] = (try? await cfb
                .from("v_cbb_input_values")
                .select("game_id, start_utc, tipoff_time_et")
                .in("game_id", values: gameIds)
                .execute()
                .value) ?? []
            var timesMap: [Int: String?] = [:]
            for row in timeRows {
                timesMap[row.gameId] = row.startUtc ?? row.tipoffTimeEt
            }
            built = built.map { game in
                NCAABGameTrendsData(
                    gameId: game.gameId,
                    gameDate: game.gameDate,
                    tipoffTime: timesMap[game.gameId] ?? nil,
                    awayTeam: game.awayTeam,
                    homeTeam: game.homeTeam,
                    awayTeamLogo: game.awayTeamLogo,
                    homeTeamLogo: game.homeTeamLogo
                )
            }
        }

        // Step 5: pre-compute sort scores per game.
        built = built.map { game in
            var copy = game
            copy.ouConsensusScore = ouConsensusStrength(for: game)
            copy.atsDominanceScore = atsDominance(for: game)
            return copy
        }

        games = sortGames(built, mode: sortMode)
        loadState = .loaded
    }

    // MARK: - Sort scoring (mirrors RN logic byte-for-byte)

    private func ouConsensusStrength(for game: NCAABGameTrendsData) -> Double {
        var total: Double = 0
        let buckets: [(awayOver: Double?, awayUnder: Double?, awayRec: String?,
                       homeOver: Double?, homeUnder: Double?, homeRec: String?)] = [
            (game.awayTeam.ouLastGameOverPct, game.awayTeam.ouLastGameUnderPct, game.awayTeam.ouLastGameRecord,
             game.homeTeam.ouLastGameOverPct, game.homeTeam.ouLastGameUnderPct, game.homeTeam.ouLastGameRecord),
            (game.awayTeam.ouFavDogOverPct, game.awayTeam.ouFavDogUnderPct, game.awayTeam.ouFavDogRecord,
             game.homeTeam.ouFavDogOverPct, game.homeTeam.ouFavDogUnderPct, game.homeTeam.ouFavDogRecord),
            (game.awayTeam.ouSideFavDogOverPct, game.awayTeam.ouSideFavDogUnderPct, game.awayTeam.ouSideFavDogRecord,
             game.homeTeam.ouSideFavDogOverPct, game.homeTeam.ouSideFavDogUnderPct, game.homeTeam.ouSideFavDogRecord),
            (game.awayTeam.ouRestBucketOverPct, game.awayTeam.ouRestBucketUnderPct, game.awayTeam.ouRestBucketRecord,
             game.homeTeam.ouRestBucketOverPct, game.homeTeam.ouRestBucketUnderPct, game.homeTeam.ouRestBucketRecord),
            (game.awayTeam.ouRestCompOverPct, game.awayTeam.ouRestCompUnderPct, game.awayTeam.ouRestCompRecord,
             game.homeTeam.ouRestCompOverPct, game.homeTeam.ouRestCompUnderPct, game.homeTeam.ouRestCompRecord)
        ]
        for b in buckets {
            let bothOver = (b.awayOver ?? 0) > minPercentage && (b.homeOver ?? 0) > minPercentage
            let bothUnder = (b.awayUnder ?? 0) > minPercentage && (b.homeUnder ?? 0) > minPercentage
            if bothOver {
                let aGames = parseNCAABRecord(b.awayRec).total
                let hGames = parseNCAABRecord(b.homeRec).total
                if aGames >= minGamesThreshold && hGames >= minGamesThreshold {
                    let totalGames = Double(aGames + hGames)
                    let avgPct = ((b.awayOver ?? 0) * Double(aGames) + (b.homeOver ?? 0) * Double(hGames)) / totalGames
                    total += avgPct * Double(min(aGames, hGames))
                }
            }
            if bothUnder {
                let aGames = parseNCAABRecord(b.awayRec).total
                let hGames = parseNCAABRecord(b.homeRec).total
                if aGames >= minGamesThreshold && hGames >= minGamesThreshold {
                    let totalGames = Double(aGames + hGames)
                    let avgPct = ((b.awayUnder ?? 0) * Double(aGames) + (b.homeUnder ?? 0) * Double(hGames)) / totalGames
                    total += avgPct * Double(min(aGames, hGames))
                }
            }
        }
        return total
    }

    private func atsDominance(for game: NCAABGameTrendsData) -> Double {
        var total: Double = 0
        let buckets: [(awayPct: Double?, awayRec: String?, homePct: Double?, homeRec: String?)] = [
            (game.awayTeam.atsLastGameCoverPct, game.awayTeam.atsLastGameRecord,
             game.homeTeam.atsLastGameCoverPct, game.homeTeam.atsLastGameRecord),
            (game.awayTeam.atsFavDogCoverPct, game.awayTeam.atsFavDogRecord,
             game.homeTeam.atsFavDogCoverPct, game.homeTeam.atsFavDogRecord),
            (game.awayTeam.atsSideFavDogCoverPct, game.awayTeam.atsSideFavDogRecord,
             game.homeTeam.atsSideFavDogCoverPct, game.homeTeam.atsSideFavDogRecord),
            (game.awayTeam.atsRestBucketCoverPct, game.awayTeam.atsRestBucketRecord,
             game.homeTeam.atsRestBucketCoverPct, game.homeTeam.atsRestBucketRecord),
            (game.awayTeam.atsRestCompCoverPct, game.awayTeam.atsRestCompRecord,
             game.homeTeam.atsRestCompCoverPct, game.homeTeam.atsRestCompRecord)
        ]
        for b in buckets {
            guard let aPct = b.awayPct, let hPct = b.homePct else { continue }
            let aGames = parseNCAABRecord(b.awayRec).total
            let hGames = parseNCAABRecord(b.homeRec).total
            let minGames = min(aGames, hGames)
            if minGames >= minGamesThreshold {
                let diff = abs(aPct - hPct)
                if diff > minATSDifference {
                    total += diff * Double(minGames)
                }
            }
        }
        return total
    }

    private func sortGames(_ list: [NCAABGameTrendsData], mode: SortMode) -> [NCAABGameTrendsData] {
        switch mode {
        case .ouConsensus:
            return list.sorted { ($0.ouConsensusScore ?? 0) > ($1.ouConsensusScore ?? 0) }
        case .atsDominance:
            return list.sorted { ($0.atsDominanceScore ?? 0) > ($1.atsDominanceScore ?? 0) }
        case .time:
            return list.sorted { a, b in
                if let aT = a.tipoffTime, let bT = b.tipoffTime { return aT < bT }
                if a.tipoffTime != nil && b.tipoffTime == nil { return true }
                if a.tipoffTime == nil && b.tipoffTime != nil { return false }
                return a.gameDate < b.gameDate
            }
        }
    }

    private struct MappingRow: Decodable, Sendable {
        let apiTeamId: Int
        let espnTeamIdInt: Int?
        let espnTeamIdString: String?

        enum CodingKeys: String, CodingKey {
            case apiTeamId = "api_team_id"
            case espnTeamId = "espn_team_id"
        }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            apiTeamId = try c.decode(Int.self, forKey: .apiTeamId)
            if let n = try? c.decode(Int.self, forKey: .espnTeamId) {
                espnTeamIdInt = n
                espnTeamIdString = nil
            } else if let s = try? c.decode(String.self, forKey: .espnTeamId) {
                espnTeamIdInt = nil
                espnTeamIdString = s
            } else {
                espnTeamIdInt = nil
                espnTeamIdString = nil
            }
        }
    }

    private struct TimeRow: Decodable, Sendable {
        let gameId: Int
        let startUtc: String?
        let tipoffTimeEt: String?
        enum CodingKeys: String, CodingKey {
            case gameId = "game_id"
            case startUtc = "start_utc"
            case tipoffTimeEt = "tipoff_time_et"
        }
    }
}

File: wagerproof-android-native/core/stores/src/main/kotlin/com/wagerproof/core/stores/NCAABBettingTrendsStore.kt
package com.wagerproof.core.stores

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.wagerproof.core.models.NCAABGameTrendsData
import com.wagerproof.core.models.NCAABSituationalTrendRow
import com.wagerproof.core.models.parseNCAABRecord
import com.wagerproof.core.models.serialization.FlexibleIntSerializer
import com.wagerproof.core.services.SupabaseClients
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.math.abs
import kotlin.math.min

/**
 * Port of iOS `NCAABBettingTrendsStore.swift`. Near-twin of
 * [NBABettingTrendsStore] — see it for the shared structure. NCAAB adds team
 * logos (joined from `ncaab_team_mapping`), pulls tipoff from
 * `v_cbb_input_values` (`start_utc ?? tipoff_time_et`), and pre-computes the
 * two sort scores onto the model (NBA computes them on demand).
 *
 * Fetches from `ncaab_game_situational_trends_today`; falls back to
 * `ncaab_game_situational_trends` when the today view is empty (matches RN).
 *
 * The O/U-consensus + ATS-dominance scoring math must produce identical scores
 * to iOS — transcribed byte-for-byte from the Swift store. NCAAB tables live on
 * the CFB/sports-data project → `SupabaseClients.cfb`.
 */
@Stable
class NCAABBettingTrendsStore {
    enum class SortMode { time, ouConsensus, atsDominance }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    var games by mutableStateOf<List<NCAABGameTrendsData>>(emptyList()); private set
    var loadState by mutableStateOf<LoadState>(LoadState.Idle); private set

    /**
     * Set on every successful refresh. Drives the TTL guard below and the game
     * sheets' first-hydrate skeleton rule (`loading && lastFetched == null`).
     * Epoch millis (matches sibling stores' `System.currentTimeMillis`).
     */
    var lastFetched by mutableStateOf<Long?>(null); private set

    private val minGamesThreshold: Int = 5
    private val minPercentage: Double = 55.0
    private val minATSDifference: Double = 10.0

    fun close() = scope.cancel()

    /** Lookup a single game by id. Mirrors RN `useNCAABBettingTrendsForGame`. */
    fun trends(forGameId: Int): NCAABGameTrendsData? =
        games.firstOrNull { it.gameId == forGameId }

    /**
     * Idempotent hydrate for sheet-local stores — skips the network while a
     * fetch is in flight or the slate is fresh. Mirrors
     * `MLBBettingTrendsStore.refreshIfNeeded`.
     */
    suspend fun refreshIfNeeded(maxAgeSeconds: Long = 600) {
        if (loadState is LoadState.Loading || loadState is LoadState.Refreshing) return
        lastFetched?.let {
            if (System.currentTimeMillis() - it < maxAgeSeconds * 1000L) return
        }
        refresh()
    }

    suspend fun refresh() {
        // FIDELITY-WAIVER #101: iOS DummyDataMode branch has no Android
        // equivalent — go straight to live.
        loadState = LoadState.Loading
        val cfb = SupabaseClients.cfb

        // Step 1: situational trends (today's table → fallback).
        val trendsRows: List<NCAABSituationalTrendRow> = try {
            val primary = runCatching {
                cfb.from("ncaab_game_situational_trends_today")
                    .select {
                        order("game_date", Order.ASCENDING)
                        order("game_id", Order.ASCENDING)
                    }
                    .decodeList<NCAABSituationalTrendRow>()
            }.getOrDefault(emptyList())
            if (primary.isNotEmpty()) {
                primary
            } else {
                cfb.from("ncaab_game_situational_trends")
                    .select {
                        order("game_date", Order.ASCENDING)
                        order("game_id", Order.ASCENDING)
                    }
                    .decodeList<NCAABSituationalTrendRow>()
            }
        } catch (e: Exception) {
            loadState = LoadState.Failed("Failed to fetch NCAAB trends: ${e.message}")
            return
        }
        if (trendsRows.isEmpty()) {
            games = emptyList()
            loadState = LoadState.Loaded
            lastFetched = System.currentTimeMillis()
            return
        }

        // Step 2: team logos.
        val mappingRows = runCatching {
            cfb.from("ncaab_team_mapping")
                .select(Columns.raw("api_team_id, espn_team_id"))
                .decodeList<MappingRow>()
        }.getOrDefault(emptyList())
        val teamLogoMap = HashMap<Int, String?>()
        for (row in mappingRows) {
            val espnId = row.espnTeamId
            val logo = espnId?.let { "https://a.espncdn.com/i/teamlogos/ncaa/500/$it.png" }
            teamLogoMap[row.apiTeamId] = logo
        }

        // Step 3: group rows by game_id using team_side.
        val partial = LinkedHashMap<Int, Partial>()
        for (row in trendsRows) {
            val side = row.teamSide.lowercase()
            if (side != "away" && side != "home") continue
            val logo = teamLogoMap[row.apiTeamId]
            val existing = partial.getOrPut(row.gameId) { Partial(gameDate = row.gameDate) }
            if (side == "away") {
                existing.away = row
                existing.awayLogo = logo
            } else {
                existing.home = row
                existing.homeLogo = logo
            }
        }

        var built: List<NCAABGameTrendsData> = partial.mapNotNull { (gameId, slots) ->
            val away = slots.away ?: return@mapNotNull null
            val home = slots.home ?: return@mapNotNull null
            NCAABGameTrendsData(
                gameId = gameId,
                gameDate = slots.gameDate,
                tipoffTime = null,
                awayTeam = away,
                homeTeam = home,
                awayTeamLogo = slots.awayLogo,
                homeTeamLogo = slots.homeLogo,
            )
        }

        // Step 4: tipoff times from v_cbb_input_values for today's game ids.
        if (built.isNotEmpty()) {
            val gameIds = built.map { it.gameId }
            val timeRows = runCatching {
                cfb.from("v_cbb_input_values")
                    .select(Columns.raw("game_id, start_utc, tipoff_time_et")) {
                        filter { isIn("game_id", gameIds) }
                    }
                    .decodeList<TimeRow>()
            }.getOrDefault(emptyList())
            val timesMap = HashMap<Int, String?>()
            for (row in timeRows) { timesMap[row.gameId] = row.startUtc ?: row.tipoffTimeEt }
            built = built.map { it.copy(tipoffTime = timesMap[it.gameId]) }
        }

        // Step 5: pre-compute sort scores per game.
        built = built.map {
            it.copy(
                ouConsensusScore = ouConsensusStrength(it),
                atsDominanceScore = atsDominance(it),
            )
        }

        games = sortGames(built, SortMode.time)
        loadState = LoadState.Loaded
        lastFetched = System.currentTimeMillis()
    }

    // MARK: - Sort scoring (mirrors RN logic byte-for-byte)

    private fun ouConsensusStrength(game: NCAABGameTrendsData): Double {
        var total = 0.0
        val a = game.awayTeam
        val h = game.homeTeam
        val buckets = listOf(
            OuBucket(a.ouLastGameOverPct, a.ouLastGameUnderPct, a.ouLastGameRecord,
                h.ouLastGameOverPct, h.ouLastGameUnderPct, h.ouLastGameRecord),
            OuBucket(a.ouFavDogOverPct, a.ouFavDogUnderPct, a.ouFavDogRecord,
                h.ouFavDogOverPct, h.ouFavDogUnderPct, h.ouFavDogRecord),
            OuBucket(a.ouSideFavDogOverPct, a.ouSideFavDogUnderPct, a.ouSideFavDogRecord,
                h.ouSideFavDogOverPct, h.ouSideFavDogUnderPct, h.ouSideFavDogRecord),
            OuBucket(a.ouRestBucketOverPct, a.ouRestBucketUnderPct, a.ouRestBucketRecord,
                h.ouRestBucketOverPct, h.ouRestBucketUnderPct, h.ouRestBucketRecord),
            OuBucket(a.ouRestCompOverPct, a.ouRestCompUnderPct, a.ouRestCompRecord,
                h.ouRestCompOverPct, h.ouRestCompUnderPct, h.ouRestCompRecord),
        )
        for (b in buckets) {
            val bothOver = (b.awayOver ?: 0.0) > minPercentage && (b.homeOver ?: 0.0) > minPercentage
            val bothUnder = (b.awayUnder ?: 0.0) > minPercentage && (b.homeUnder ?: 0.0) > minPercentage
            if (bothOver) {
                val aGames = parseNCAABRecord(b.awayRec).total
                val hGames = parseNCAABRecord(b.homeRec).total
                if (aGames >= minGamesThreshold && hGames >= minGamesThreshold) {
                    val totalGames = (aGames + hGames).toDouble()
                    val avgPct = ((b.awayOver ?: 0.0) * aGames + (b.homeOver ?: 0.0) * hGames) / totalGames
                    total += avgPct * min(aGames, hGames)
                }
            }
            if (bothUnder) {
                val aGames = parseNCAABRecord(b.awayRec).total
                val hGames = parseNCAABRecord(b.homeRec).total
                if (aGames >= minGamesThreshold && hGames >= minGamesThreshold) {
                    val totalGames = (aGames + hGames).toDouble()
                    val avgPct = ((b.awayUnder ?: 0.0) * aGames + (b.homeUnder ?: 0.0) * hGames) / totalGames
                    total += avgPct * min(aGames, hGames)
                }
            }
        }
        return total
    }

    private fun atsDominance(game: NCAABGameTrendsData): Double {
        var total = 0.0
        val a = game.awayTeam
        val h = game.homeTeam
        val buckets = listOf(
            AtsBucket(a.atsLastGameCoverPct, a.atsLastGameRecord, h.atsLastGameCoverPct, h.atsLastGameRecord),
            AtsBucket(a.atsFavDogCoverPct, a.atsFavDogRecord, h.atsFavDogCoverPct, h.atsFavDogRecord),
            AtsBucket(a.atsSideFavDogCoverPct, a.atsSideFavDogRecord, h.atsSideFavDogCoverPct, h.atsSideFavDogRecord),
            AtsBucket(a.atsRestBucketCoverPct, a.atsRestBucketRecord, h.atsRestBucketCoverPct, h.atsRestBucketRecord),
            AtsBucket(a.atsRestCompCoverPct, a.atsRestCompRecord, h.atsRestCompCoverPct, h.atsRestCompRecord),
        )
        for (b in buckets) {
            val aPct = b.awayPct ?: continue
            val hPct = b.homePct ?: continue
            val aGames = parseNCAABRecord(b.awayRec).total
            val hGames = parseNCAABRecord(b.homeRec).total
            val minGames = min(aGames, hGames)
            if (minGames >= minGamesThreshold) {
                val diff = abs(aPct - hPct)
                if (diff > minATSDifference) {
                    total += diff * minGames
                }
            }
        }
        return total
    }

    private fun sortGames(list: List<NCAABGameTrendsData>, mode: SortMode): List<NCAABGameTrendsData> =
        when (mode) {
            SortMode.ouConsensus -> list.sortedByDescending { it.ouConsensusScore ?: 0.0 }
            SortMode.atsDominance -> list.sortedByDescending { it.atsDominanceScore ?: 0.0 }
            SortMode.time -> list.sortedWith(TIME_COMPARATOR)
        }

    private class Partial(
        var away: NCAABSituationalTrendRow? = null,
        var home: NCAABSituationalTrendRow? = null,
        var awayLogo: String? = null,
        var homeLogo: String? = null,
        val gameDate: String,
    )

    private data class OuBucket(
        val awayOver: Double?, val awayUnder: Double?, val awayRec: String?,
        val homeOver: Double?, val homeUnder: Double?, val homeRec: String?,
    )

    private data class AtsBucket(
        val awayPct: Double?, val awayRec: String?, val homePct: Double?, val homeRec: String?,
    )

    @Serializable
    private data class MappingRow(
        @SerialName("api_team_id") val apiTeamId: Int,
        @Serializable(with = FlexibleIntSerializer::class)
        @SerialName("espn_team_id") val espnTeamId: Int? = null,
    )

    @Serializable
    private data class TimeRow(
        @SerialName("game_id") val gameId: Int,
        @SerialName("start_utc") val startUtc: String? = null,
        @SerialName("tipoff_time_et") val tipoffTimeEt: String? = null,
    )

    companion object {
        /** some-tipoff-before-none, else compare tipoff string, else gameDate. */
        private val TIME_COMPARATOR = Comparator<NCAABGameTrendsData> { a, b ->
            val aT = a.tipoffTime
            val bT = b.tipoffTime
            when {
                aT != null && bT != null -> aT.compareTo(bT)
                aT != null && bT == null -> -1
                aT == null && bT != null -> 1
                else -> a.gameDate.compareTo(b.gameDate)
            }
        }
    }
}
